substrate/frame/sassafras/src/lib.rs

```diff
@@ -0,0 +1,1040 @@
+		/// Phantom config
+		#[serde(skip)]
+		pub _phantom: sp_std::marker::PhantomData<T>,
+	}
+
+	#[pallet::genesis_build]
+	impl<T: Config> BuildGenesisConfig for GenesisConfig<T> {
+		fn build(&self) {
+			EpochConfig::<T>::put(self.epoch_config);
+			Pallet::<T>::initialize_genesis_authorities(&self.authorities);
+
+			#[cfg(feature = "construct-dummy-ring-context")]
+			{
+				debug!(target: LOG_TARGET, "Constructing dummy ring context");
+				let ring_ctx = vrf::RingContext::new_testing();
+				RingContext::<T>::put(ring_ctx);
+				Pallet::<T>::update_ring_verifier(&self.authorities);
+			}
+		}
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		/// Block initialization.
+		fn on_initialize(now: BlockNumberFor<T>) -> Weight {
+			let claim = <frame_system::Pallet<T>>::digest()
+				.logs
+				.iter()
+				.filter_map(|item| item.pre_runtime_try_to::<SlotClaim>(&SASSAFRAS_ENGINE_ID))
+				.next()
+				.expect("Valid block must have a slot claim. qed");
+
+			CurrentSlot::<T>::put(claim.slot);
+
+			// As the slots may not be zero-based, we need to keep track of what is the
+			// slot used fo the first block.
+			if <frame_system::Pallet<T>>::block_number() == One::one() {
+				GenesisSlot::<T>::put(claim.slot);
+
+				// Deposit a log as this is the first block in first epoch.
+				let next_epoch = NextEpochDescriptor {
+					randomness: Self::next_randomness(),
+					authorities: Self::next_authorities().into_inner(),
+					config: None,
+				};
+				Self::deposit_next_epoch_descriptor_digest(next_epoch);
+			}
+
+			let randomness_output = claim
+				.vrf_signature
+				.outputs
+				.get(0)
+				.expect("Presence should have been already checked by the client; qed");
+			ClaimTemporaryData::<T>::put(randomness_output);
+
+			// Enact epoch change, if necessary.
+			T::EpochChangeTrigger::trigger::<T>(now);
+
+			Weight::zero()
+		}
+
+		/// Block finalization.
+		fn on_finalize(_now: BlockNumberFor<T>) {
+			// At the end of the block, we can safely include the current slot randomness
+			// to the randomness accumulator. If we've determined that this block was the
+			// first in a new epoch, the changeover logic has already occurred at this point
+			// (i.e. `enact_epoch_change` has already been called).
+			let randomness_input = vrf::slot_claim_input(
+				&Self::randomness(),
+				CurrentSlot::<T>::get(),
+				EpochIndex::<T>::get(),
+			);
+			let randomness_output = ClaimTemporaryData::<T>::take()
+				.expect("Finalization is called after initialization; qed");
+			let randomness = randomness_output
+				.make_bytes::<RANDOMNESS_LENGTH>(RANDOMNESS_VRF_CONTEXT, &randomness_input);
+			Self::deposit_slot_randomness(&randomness);
+
+			// Check if we are in the epoch's second half.
+			// If so, start sorting the next epoch tickets.
+			let epoch_length = T::EpochLength::get();
+			let current_slot_idx = Self::current_slot_index();
+			if current_slot_idx >= epoch_length / 2 {
```

Comment:
Random thought: For this sorting in the second half of the epoch, did you think about leveraging `on_idle` to get some extra work done?

Reply:
This is a very good idea which also reminds me that maybe we can even do better.

At the moment we wait for the second half of the epoch to start sorting the tickets. There is not point doing this.
We can start sorting immediately in the `on_idle` hook.

Sorting in `on_finalize` apparently can be removed completely.

The only catch is that I have limit the number of sorted segments from the `on_idle` remaining weight

Reply:
> The only catch is that I have limit the number of sorted segments from the on_idle remaining weight

Yeah, when we are busy and there is not a lot remaining weight left, in the worst case we would always sort when the new epoch is enacted. I don't know how bad that is.

Reply:
I think is fine. 

I we leave it in on_finalize (as it is now) we always silently consume computing power without never accounting for it.

Also, from the benchmarks, sorting the segments is fast.

Lets keep open the discussion BTW

Reply:
> If we leave it in on_finalize (as it is now) we always silently consume resources (Weight) without never accounting for it.

All in favor of using `on_idle`, but it could be accounted for. Afaik `on_initialize` should also return the weight of `on_finalize`. So in the second half of the epoch you could estimate how much work will be done.

Reply:
Yeah `on_initialize` weight must account for on finalize weight as well.

https://github.com/paritytech/polkadot-sdk/blob/4ddc48d8654257b503acf05afbdae78c66f3d502/substrate/frame/sassafras/src/benchmarking.rs#L68-L71

---

I think we need a small brainstorm to sort out if is better to fully delegate to `on_idle` or to do that also `on_finalize`.

In both case I need to account for the Weight to partially sort X segments